### PR TITLE
[Impeller] Add a safe way for runtime effects to provide dynamically generated shader metadata

### PR DESCRIPTION
**Pull request by jason-simmons**
See https://github.com/flutter/flutter/issues/113715

### CODE REVIEW

File: impeller/entity/contents/runtime_effect_contents.cc [hunk 1]
```diff
@@ -165,7 +165,8 @@ bool RuntimeEffectContents::Render(const ContentContext& renderer,
   for (auto uniform : runtime_stage_->GetUniforms()) {
     // TODO(113715): Populate this metadata once GLES is able to handle
```
Comment (chinmaygarde) [inline impeller/entity/contents/runtime_effect_contents.cc:166]: This comment can now be removed.

File: impeller/renderer/command.h [hunk 2]
```diff
@@ -28,13 +28,26 @@ namespace impeller {
...
 
-  Resource(const ShaderMetadata* p_isa, ResourceType p_resource)
-      : isa(p_isa), resource(p_resource) {}
+  Resource(const ShaderMetadata* isa, ResourceType p_resource)
+      : resource(p_resource), isa_(isa) {}
+
+  Resource(std::shared_ptr<const ShaderMetadata>& isa, ResourceType p_resource)
+      : resource(p_resource), isa_(nullptr), dynamic_isa_(isa) {}
+
+  const ShaderMetadata* isa() const {
```
Comment (chinmaygarde) [inline impeller/renderer/command.h:41]: Perhaps just call this `GetMetadata` and change the ivar to `metadata_`. Admittedly, the `isa` ptr is a holdover from the Objective-C runtime and its not particularly descriptive if you haven't done some runtime mangling.

Comment (chinmaygarde) [reply impeller/renderer/command.h:41]: I like the technique though.

File: impeller/renderer/command.h [hunk 3]
```diff
@@ -28,13 +28,26 @@ namespace impeller {
...
+    return dynamic_isa_ ? dynamic_isa_.get() : isa_;
+  }
+
+ private:
+  // Static shader metadata (typically generated by ImpellerC).
+  const ShaderMetadata* isa_;
```
Comment (chinmaygarde) [inline impeller/renderer/command.h:47]: Default it to `nullptr` here in case we forget to explicitly do so in the ctor now or if more are added.

**Review by chinmaygarde — APPROVED** [review]
Some nits but otherwise lgtm.

File: impeller/entity/contents/runtime_effect_contents.cc (hunk 1 above)
Comment (jason-simmons) [reply impeller/entity/contents/runtime_effect_contents.cc:166]: Keeping the comment - the `ShaderMetadata` is still not actually being filled with the data required to use runtime effects on GLES

File: impeller/renderer/command.h (hunk 2 above)
Comment (jason-simmons) [reply impeller/renderer/command.h]: done

File: impeller/renderer/command.h (hunk 3 above)
Comment (jason-simmons) [reply impeller/renderer/command.h]: done

**Pull request merged**